File: backend/src/places/places.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

export interface PlaceResult {
  placeId: string;
  name: string;
  formattedAddress: string;
  location: { lat: number; lng: number };
  photoRef: string | null;
  rating: number | null;
  types: string[];
}

@Injectable()
export class PlacesService {
  private readonly logger = new Logger(PlacesService.name);
  private readonly apiKey: string;

  constructor(private readonly configService: ConfigService) {
    this.apiKey = this.configService.get<string>('GOOGLE_PLACES_API_KEY', '');
    if (!this.apiKey) {
      this.logger.warn('GOOGLE_PLACES_API_KEY not set — Places features will be unavailable');
    }
  }

  async textSearch(query: string, maxResults: number = 5): Promise<PlaceResult[]> {
    if (!this.apiKey) {
      throw new Error('Google Places API key not configured');
    }

    const url = 'https://places.googleapis.com/v1/places:searchText';
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Goog-Api-Key': this.apiKey,
        'X-Goog-FieldMask': 'places.id,places.displayName,places.formattedAddress,places.location,places.photos,places.rating,places.types',
      },
      body: JSON.stringify({
        textQuery: query,
        maxResultCount: maxResults,
        languageCode: 'ko',
      }),
    });

    if (!response.ok) {
      const errText = await response.text();
      this.logger.error(`Places API error: ${response.status} ${errText}`);
      throw new Error(`Places API error: ${response.status}`);
    }

    const data = (await response.json()) as {
      places?: Array<{
        id: string;
        displayName: { text: string };
        formattedAddress: string;
        location: { latitude: number; longitude: number };
        photos?: Array<{ name: string }>;
        rating?: number;
        types?: string[];
      }>;
    };

    return (data.places || []).map((place) => ({
      placeId: place.id,
      name: place.displayName?.text || '',
      formattedAddress: place.formattedAddress || '',
      location: {
        lat: place.location?.latitude || 0,
        lng: place.location?.longitude || 0,
      },
      photoRef: place.photos?.[0]?.name || null,
      rating: place.rating || null,
      types: place.types || [],
    }));
  }

  getPhotoUrl(photoRef: string, maxWidth: number = 800): string {
    if (!this.apiKey) return '';
    return `https://places.googleapis.com/v1/${photoRef}/media?maxWidthPx=${maxWidth}&key=${this.apiKey}`;
  }
}

File: nui/TravelAgent/Services/PlacesService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using TravelAgent.Models;

namespace TravelAgent.Services
{
    public static class PlacesService
    {
        private static readonly HttpClient _client = new HttpClient();

        private static readonly JsonSerializerOptions _jsonOptions = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true
        };

        /// <summary>
        /// Performs a text search using the Google Places API (New).
        /// </summary>
        public static async Task<List<PlaceResult>> TextSearch(string query, int maxResults = 5)
        {
            try
            {
                var url = "https://places.googleapis.com/v1/places:searchText";

                var requestBody = new
                {
                    textQuery = query,
                    languageCode = "ko",
                    maxResultCount = maxResults
                };

                var json = JsonSerializer.Serialize(requestBody);
                var request = new HttpRequestMessage(HttpMethod.Post, url)
                {
                    Content = new StringContent(json, Encoding.UTF8, "application/json")
                };

                request.Headers.Add("X-Goog-Api-Key", ApiConfig.GooglePlacesApiKey);
                request.Headers.Add("X-Goog-FieldMask",
                    "places.id,places.displayName,places.formattedAddress,places.rating,places.photos");

                var response = await _client.SendAsync(request);
                response.EnsureSuccessStatusCode();

                var responseJson = await response.Content.ReadAsStringAsync();
                var searchResponse = JsonSerializer.Deserialize<PlaceSearchResponse>(responseJson, _jsonOptions);

                return searchResponse?.Places ?? new List<PlaceResult>();
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"[PlacesService] TextSearch error: {ex.Message}");
                return new List<PlaceResult>();
            }
        }

        /// <summary>
        /// Builds the photo URL for a given Places photo resource name.
        /// </summary>
        public static string GetPhotoUrl(string photoName, int maxWidth = 800)
        {
            if (string.IsNullOrEmpty(photoName))
                return "";

            return $"https://places.googleapis.com/v1/{photoName}/media?maxWidthPx={maxWidth}&key={ApiConfig.GooglePlacesApiKey}";
        }
    }
}
